pkg/providers/cloudstack/cloudstack.go

```diff
@@ -1003,6 +1006,7 @@ func (p *cloudstackProvider) generateCAPISpecForUpgrade(ctx context.Context, boo
 	if err != nil {
 		return nil, nil, err
 	}
+	currentSpec.CloudStackDatacenter = csdc
```

Comment:
why are we modifying the cluster spec here? This should be set during the object construction. This creates a dangerous side effect/invariant

Reply:
From what I can tell, currentSpec's plumbing into this method can be traced back through

* generateCAPISpecForUpgrade (cloudstack.go)
* GenerateCAPISpecForUpgrade (cloudstack.go)
* clusterManager calling GetCurrentClusterSpec (cluster_manager.go)

So the cluster spec is originally constructed in GetCurrentClusterSpec. However, there is no logic anywhere to account for setting the provider-specific specs, like spec.CloudStackDatacenter, which resulted in currentSpec.CloudStackDatacenter being `nil`. I'm not sure where else this logic could exist, and without persisting CloudStackDatacenterConfig in the provider, we need to have currentSpec.CloudStackDatacenter set somewhere.

Reply:
We used to pass the cloudstack datacenter config (aka. csdc) to other functions as a separate parameter. However, as per Joey's suggestion I removed it because it is a part of cloudstack cluster spec (aka. spec). During the upgrade test I found the spec had a nil for its csdc so I set it to avoid nil pointer dereference error. If we don't want to do this we should go back to the original function prototypes where csdc is explicitly passed in.

Reply:
I see. I think we just need to modify the code that builds a clusterSpec from the CRs in the cluster (what's called here `currentSpec`) and make sure it pulls the cloudstack datacenterconfig and set it in the `cluster.Config`

Reply:
>  I think we just need to modify the code that builds a clusterSpec from the CRs

There doesn't appear to be any provider specific logic in [the code that builds the clusterSpecs](https://github.com/aws/eks-anywhere/blob/23b12d10a9614b62edc86720f587c09f32ca0149/pkg/clustermanager/cluster_manager.go#L1038) which would enable setting this field. Are you suggesting we modify this function somehow? Otherwise it seems like the provider-specific implementation should happen here.

Reply:
Created https://github.com/aws/eks-anywhere/issues/2929 per Joey's advice